src/shims/io.rs

```diff
@@ -133,28 +154,63 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         let count = this.read_scalar(count_op)?.to_usize(&*this.tcx)?;
 
         // Remove the file handle to avoid borrowing issues
-        this.remove_handle_and(
-            fd,
-            |mut handle, this| {
-                let bytes = handle
-                    .file
-                    .read(this.memory_mut().get_mut(buf.alloc_id)?.get_bytes_mut(
-                        tcx,
-                        buf,
-                        Size::from_bytes(count),
-                    )?)
-                    .map(|bytes| bytes as i64);
-                // Reinsert the file handle
-                this.machine.file_handler.handles.insert(fd, handle);
-                this.consume_result(bytes)
-            },
-        )
+        this.remove_handle_and(fd, |mut handle, this| {
+            // Don't use `?` to avoid returning before reinserting the handle
+
+            let bytes = this.memory_mut().get_mut(buf.alloc_id).and_then(|alloc| {
+                if count != 0 {
+                    alloc
+                        .get_bytes_mut(tcx, buf, Size::from_bytes(count))
+                        .map(|buffer| handle.file.read(buffer))
+                } else {
+                    // Using `Allocation::get_bytes_mut` requires to actually put data into
+                    // the allocation even if we take 0 bytes of it. In that case we use an
+                    // empty buffer instead to avoid dangling pointers.
+                    let mut empty = Vec::new();
+                    let result = Ok(handle.file.read(empty.as_mut_slice()));
```

Comment:
You should be able to use `&mut []`

Reply:
In fact, can't you just return `OK(0)` here and not touch file at all?

Reply:
Thank you! I don't know why, but if I do this, I still get the dangling pointer error.

Reply:
Uhm. .. do you have a backtrace? This seems extremely suspicious. I also wonder why there needs to be an OK around the read (doesn't the read return a result?)

Reply:
yeah but returns an `std::io::Result` and using `and_then` requires that the results are the same, so I had to use `map` instead.

Reply:
```
An error occurred in miri:
stack backtrace:
   0: <rustc::mir::interpret::error::InterpErrorInfo as core::convert::From<rustc::mir::interpret::error::InterpError>>::from
   1: <T as core::convert::Into<U>>::into
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/libcore/convert.rs:547
      miri::intptrcast::GlobalState::int_to_ptr
             at src/intptrcast.rs:58
   2: <miri::machine::Evaluator as rustc_mir::interpret::machine::Machine>::int_to_ptr
             at src/machine.rs:354
      rustc_mir::interpret::memory::Memory<M>::force_ptr
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc_mir/interpret/memory.rs:894
   3: rustc_mir::interpret::eval_context::InterpCx<M>::force_ptr
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc_mir/interpret/eval_context.rs:229
      miri::shims::io::EvalContextExt::read
             at src/shims/io.rs:153
      miri::shims::foreign_items::EvalContextExt::emulate_foreign_item
             at src/shims/foreign_items.rs:465
   4: miri::shims::EvalContextExt::find_fn
             at src/shims/mod.rs:48
      <miri::machine::Evaluator as rustc_mir::interpret::machine::Machine>::find_fn
             at src/machine.rs:168
      rustc_mir::interpret::terminator::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::eval_fn_call
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc_mir/interpret/terminator.rs:296
   5: rustc_mir::interpret::terminator::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::eval_terminator
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc_mir/interpret/terminator.rs:94
      rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::terminator
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc_mir/interpret/step.rs:293
      rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::step
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc_mir/interpret/step.rs:69
   6: rustc_mir::interpret::step::<impl rustc_mir::interpret::eval_context::InterpCx<M>>::run
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc_mir/interpret/step.rs:40
      miri::eval::eval_main::{{closure}}
             at src/eval.rs:191
      miri::eval::eval_main
             at src/eval.rs:190
   7: <miri::MiriCompilerCalls as rustc_driver::Callbacks>::after_analysis::{{closure}}
             at src/bin/miri.rs:52
      rustc_interface::passes::BoxedGlobalCtxt::enter::{{closure}}::{{closure}}
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc_interface/passes.rs:808
      rustc::ty::context::tls::enter_global::{{closure}}
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc/ty/context.rs:1872
      rustc::ty::context::tls::enter_context::{{closure}}
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc/ty/context.rs:1840
      rustc::ty::context::tls::set_tlv
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc/ty/context.rs:1773
      rustc::ty::context::tls::enter_context
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc/ty/context.rs:1839
      rustc::ty::context::tls::enter_global
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc/ty/context.rs:1871
      rustc_interface::passes::BoxedGlobalCtxt::enter::{{closure}}
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc_interface/passes.rs:808
      rustc_interface::passes::BoxedGlobalCtxt::access::{{closure}}
             at <::rustc_data_structures::box_region::declare_box_region_type macros>:21
   8: rustc_interface::passes::create_global_ctxt::{{closure}}
   9: alloc::boxed::<impl core::ops::generator::Generator for core::pin::Pin<alloc::boxed::Box<G>>>::resume
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/liballoc/boxed.rs:1073
      rustc_data_structures::box_region::PinnedGenerator<I,A,R>::access
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc_data_structures/box_region.rs:52
      rustc_interface::passes::BoxedGlobalCtxt::access
             at <::rustc_data_structures::box_region::declare_box_region_type macros>:24
      rustc_interface::passes::BoxedGlobalCtxt::enter
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/librustc_interface/passes.rs:808
      <miri::MiriCompilerCalls as rustc_driver::Callbacks>::after_analysis
             at src/bin/miri.rs:45
  10: rustc_interface::interface::run_compiler_in_existing_thread_pool
  11: std::thread::local::LocalKey<T>::with
  12: syntax::with_globals
  13: std::sys_common::backtrace::__rust_begin_short_backtrace
  14: __rust_maybe_catch_panic
             at src/libpanic_unwind/lib.rs:80
  15: core::ops::function::FnOnce::call_once{{vtable.shim}}
  16: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/liballoc/boxed.rs:922
  17: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
             at /rustc/084beb83e0e87d673d5fabc844d28e8e8ae2ab4c/src/liballoc/boxed.rs:922
      std::sys_common::thread::start_thread
             at src/libstd/sys_common/thread.rs:13
      std::sys::unix::thread::Thread::new::thread_start
             at src/libstd/sys/unix/thread.rs:79
  18: start_thread
  19: __clone
```

Reply:
Should I open an issue for this?